lib/streamlit/elements/arrow_table.py

```diff
@@ -0,0 +1,165 @@
+    css_rules = []
+
+    if "table_styles" in styles:
+        table_styles = styles["table_styles"]
+        table_styles = _trim_pandas_styles(table_styles)
+        for style in table_styles:
+            # NB: styles in "table_styles" have a space
+            # between the UUID and the selector.
+            rule = _pandas_style_to_css(style, styler.uuid, separator=" ")
+            css_rules.append(rule)
+
+    if "cellstyle" in styles:
+        cellstyle = styles["cellstyle"]
+        cellstyle = _trim_pandas_styles(cellstyle)
+        for style in cellstyle:
+            rule = _pandas_style_to_css(style, styler.uuid)
+            css_rules.append(rule)
+
+    proto.styles = "\n".join(css_rules)
+
+
+def _trim_pandas_styles(styles):
+    # Filter out empty styles, as every cell will have a class
+    # but the list of props may just be [['', '']].
+    return [x for x in styles if any(any(y) for y in x["props"])]
+
+
+def _pandas_style_to_css(style, uuid, separator=""):
+    from collections import OrderedDict
+
+    declarations = []
+    for css_property, css_value in style["props"]:
+        declaration = css_property.strip() + ": " + css_value.strip()
+        declarations.append(declaration)
+
+    # Remove duplicate styles preserving the order and
+    # keeping the first appearance of each element.
+    declarations = list(OrderedDict.fromkeys(declarations))
+
+    table_selector = "#T_" + str(uuid)
+    cell_selector = style["selector"]
+    selector = table_selector + separator + cell_selector
+    declaration_block = "; ".join(declarations)
+    rule_set = selector + " { " + declaration_block + " }"
+
+    return rule_set
+
+
+def _marshall_headers(proto, df):
+    # Serialize header columns
+    header_columns = map(util._tuple_to_list, df.columns.values)
+    header_columns_df = pd.DataFrame(header_columns)
+    header_columns_table = _dataframe_to_serialized_arrow_table(header_columns_df)
+    proto.header_columns = header_columns_table
+
+    # Serialize header rows
+    header_rows = map(util._tuple_to_list, df.index.values)
+    header_rows_df = pd.DataFrame(header_rows)
+    header_rows_table = _dataframe_to_serialized_arrow_table(header_rows_df)
+    proto.header_rows = header_rows_table
+
+
+def _dataframe_to_serialized_arrow_table(df):
+    table = pa.Table.from_pandas(df)
+    sink = pa.BufferOutputStream()
+    writer = pa.RecordBatchStreamWriter(sink, table.schema)
+    writer.write_table(table)
+    writer.close()
+    return sink.getvalue().to_pybytes()
+
+
+def _marshall_data(proto, df):
+    # Reset row and column indices
+    df = df.T.reset_index(drop=True)
+    df = df.T.reset_index(drop=True)
+
+    # Serialize data
+    proto.data = _dataframe_to_serialized_arrow_table(df)
+
+
+def _use_display_values(df, styles):
+    # TODO: rewrite this method without using regex
+    import re
+
+    # Convert all dataframe values to strings as we
+    # only use them to show the result.
```

Comment:
This is for the styler only, right? We pass the data as binary, don't we.

Reply:
Yes. I'm creating a new dataframe that uses display values instead of the original ones, converting that to binary, and then passing it over the proto.